Fix bug in ShardedTensorMetadata serde.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63902

The 'memory_format' field was not being serialized correctly and used
the same encoding for different fields.

Differential Revision: [D30527324](https://our.internmc.facebook.com/intern/diff/D30527324/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

pritamdamania committed:
Fix bug in ShardedTensorMetadata serde.

The 'memory_format' field was not being serialized correctly and used
the same encoding for different fields.

Differential Revision: [D30527324](https://our.internmc.facebook.com/intern/diff/D30527324/)

[ghstack-poisoned]

bowangbj commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -122,6 +124,52 @@ def wrapper(self):
+        layouts = [torch.strided, torch.sparse_coo]
+        requires_grad = [True, False]
+        memory_format = [torch.contiguous_format, torch.channels_last, torch.preserve_format]
+        pin_memory = [True, False]
```
nit:  require_grads  memory_formats  pin_memories

bowangbj commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -122,6 +124,52 @@ def wrapper(self):
+            dtype, layout, requires_grad, memory_format, pin_memory = test
+
+            expected_st_metadata = _sharded_tensor.ShardedTensorMetadata(
+                shards_metadata, (10, 10), dtype, layout, requires_grad,
```
maybe use tensor_properties ?

bowangbj commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -122,6 +124,52 @@ def wrapper(self):
+        memory_format = [torch.contiguous_format, torch.channels_last, torch.preserve_format]
+        pin_memory = [True, False]
+
+        for test in itertools.product(dtypes, layouts, requires_grad, memory_format, pin_memory):
```
nit: 
s/test/tensor_properties_input

bowangbj commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -122,6 +124,52 @@ def wrapper(self):
+class TestShardedTensorMetadata(TestCase):
+    def test_serde(self):
+        shards_metadata = [
+            ShardMetadata(
```
nit: shard_metadatas ?

bowangbj commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -122,6 +124,52 @@ def wrapper(self):
     return wrapper
 
+class TestShardedTensorMetadata(TestCase):
+    def test_serde(self):
```
probably rename it to 

test_serialize_and_deserialize 

for better readability ?

bowangbj commented:
Thanks Pritam for the great fix, just a few small nits.

bowangbj reviewed (commented):
Thanks Pritam for the great fix, just a few small nits.

SciPioneer commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -85,7 +85,7 @@ def __getstate__(self):
             mem_format_encoding = 1
         elif self.memory_format == torch.preserve_format:
-            mem_format_encoding = 1
+            mem_format_encoding = 2
```
Shall we replace the numeric values here with more readable enum values?

pritamdamania committed:
Update on "Fix bug in ShardedTensorMetadata serde."


The 'memory_format' field was not being serialized correctly and used
the same encoding for different fields.

Differential Revision: [D30527324](https://our.internmc.facebook.com/intern/diff/D30527324/)

[ghstack-poisoned]

pritamdamania87 commented:
@SciPioneer @bowangbj Can you take another look at this PR? Thanks!

bowangbj reviewed (approved):
Looks great, thanks Pritam!

pritamdamania committed:
Update on "Fix bug in ShardedTensorMetadata serde."


The 'memory_format' field was not being serialized correctly and used
the same encoding for different fields.

Differential Revision: [D30527324](https://our.internmc.facebook.com/intern/diff/D30527324/)

[ghstack-poisoned]
